overlay.go

```diff
@@ -0,0 +1,346 @@
+	for _, key := range val.MapKeys() {
+		v := val.MapIndex(key)
+		if !v.CanInterface() {
+			continue
+		}
+
+		if ctx.indexedElemPreVisitor != nil {
+			ctx.indexedElemPreVisitor(key)
+		}
+
+		result = visitField(ctx, v.Interface()) || result
+
+		if ctx.indexedElemPostVisitor != nil {
+			ctx.indexedElemPostVisitor(key)
+		}
+	}
+
+	return result
+}
+
+// visitFieldsInSlice iterates the slice specified by val and recursively
+// visits each element. The returned value is the logical OR of the responses
+// returned by visiting all slice elements.
+func visitFieldsInSlice(ctx *visitorContext, val reflect.Value) (result bool) {
+	for i := 0; i < val.Len(); i++ {
+		v := val.Index(i)
+		if !v.CanInterface() {
+			continue
+		}
+
+		if ctx.indexedElemPreVisitor != nil {
+			ctx.indexedElemPreVisitor(i)
+		}
+
+		result = visitField(ctx, v.Interface()) || result
+
+		if ctx.indexedElemPostVisitor != nil {
+			ctx.indexedElemPostVisitor(i)
+		}
+	}
+
+	return result
+}
+
+// clearOverlayFields is an implementation of structVisitor. It recursively
+// visits all fields in the val struct and sets the ones that are tagged as
+// overlay-only to the zero value for their particular type.
+func clearOverlayFields(ctx *visitorContext, val reflect.Value, typ reflect.Type) (retainAncestors bool) {
+	for i := 0; i < typ.NumField(); i++ {
+		structField := typ.Field(i)
+
+		// Skip non-exportable fields
+		v := val.Field(i)
+		if !v.CanInterface() {
+			continue
+		}
+
+		// No need to recurse further down; just erase the field
+		if isOverlayField(structField) {
+			v.Set(reflect.Zero(v.Type()))
+			continue
+		}
+
+		_ = visitField(ctx, v.Interface())
+		retainAncestors = true
+	}
+	return retainAncestors
+}
+
+// clearNonOverlayFields is an implementation of structVisitor. It recursively
+// visits all fields in the val struct and sets any field that does not contain
+// any overlay-only descendants to the zero value for its particular type.
+func clearNonOverlayFields(ctx *visitorContext, val reflect.Value, typ reflect.Type) (retainAncestors bool) {
+	for i := 0; i < typ.NumField(); i++ {
+		structField := typ.Field(i)
+
+		// Skip non-exportable fields
+		v := val.Field(i)
+		if !v.CanInterface() {
+			continue
+		}
+
+		// If this is an overlay field we need to preserve it and all
+		// its ancestor fields up to the root. However, we still need
+		// to visit its descendants in case we need to clear additional
+		// non-overlay fields further down the tree.
+		isOverlayField := isOverlayField(structField)
+		if isOverlayField {
+			retainAncestors = true
+		}
+
+		if retain := visitField(ctx, v.Interface()); !isOverlayField && !retain {
+			v.Set(reflect.Zero(v.Type()))
+			continue
+		}
+
+		retainAncestors = true
+	}
+	return retainAncestors
+}
+
+// isOverlayField returns true if a struct field is tagged as overlay-only.
+func isOverlayField(structField reflect.StructField) bool {
+	return structField.Tag.Get("overlay-only") == "true"
```

Comment:
offhand this feels more like it should be something like:
source: overlay-only
source: either
source: base-only
or something along those lines.
'overlay-only: true' doesn't quite feel like the right booleans that we would want to support.

What do you think? I know this is bike-shedding but "bundle-validity" "bundle-source" "allowed-in: overlay" ?

Reply:
I am really bad at naming stuff :D. I kinda like the `source: overlay-only` bit so I can make a second pass to s/static/base and rename the tag as well.